include/ppx/metrics.h

```diff
@@ -0,0 +1,198 @@
+
+#define METRICS_NO_COPY(type)   \
+    type(type&)       = delete; \
+    type(const type&) = delete; \
+    type& operator=(const type&) = delete;
+
+////////////////////////////////////////////////////////////////////////////////
+
+enum class MetricInterpretation
+{
+    NONE,
+    HIGHER_IS_BETTER,
+    LOWER_IS_BETTER,
+};
+
+struct Range
+{
+    double lowerBound = std::numeric_limits<double>::min();
+    double upperBound = std::numeric_limits<double>::max();
+};
+
+struct MetricMetadata
+{
+    std::string          name;
+    std::string          unit;
+    MetricInterpretation interpretation = MetricInterpretation::NONE;
+    Range                expectedRange;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+enum class MetricType
+{
+    GAUGE,
+    COUNTER,
+};
+
+class Metric
+{
+    friend class Run;
+
+public:
+    const MetricMetadata& GetMetadata() const;
+    MetricType            GetType() const;
+
+protected:
+    Metric(const MetricMetadata& metadata, MetricType type);
+    virtual ~Metric();
+
+private:
+    MetricMetadata mMetadata;
+    MetricType     mType;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+struct GaugeStatistics
+{
+    double min;
+    double max;
+    double average;
+    double median;
+    double standardDeviation;
+    double timeRatio;
+    double percentile90;
+    double percentile95;
+    double percentile99;
+};
+
+class MetricGauge final : public Metric
+{
+public:
+    MetricGauge(const MetricMetadata& metadata);
+
+    void   RecordEntry(double seconds, double value);
+    size_t GetEntriesCount() const;
+    void   GetEntry(size_t index, double& seconds, double& value) const;
+
+    const GaugeStatistics GetStatistics(bool realtime) const;
+
+private:
+    struct TimeSeriesEntry
+    {
+        double seconds;
+        double value;
+    };
+
+private:
+    ~MetricGauge();
+    METRICS_NO_COPY(MetricGauge)
+
+    void UpdateRealTimeStatistics(double seconds, double value);
+
+private:
+    std::vector<TimeSeriesEntry> mTimeSeries;
+    GaugeStatistics              mRealTimeStatistics;
+    double                       mAccumlatedValue;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+class MetricCounter final : public Metric
+{
+    friend class Run;
+
+public:
+    MetricCounter(const MetricMetadata& metadata);
+    uint64_t Increment(uint64_t add);
+    uint64_t Get() const;
+
+private:
+    ~MetricCounter();
+    METRICS_NO_COPY(MetricCounter)
+
+private:
+    uint64_t mCounter;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+class Run final
+{
+    friend class Manager;
+
+public:
+    Result       AddMetricGauge(MetricGauge*& outMetric, MetricMetadata metadata);
+    MetricGauge* GetMetricGauge(const char* name) const;
+
+    Result         AddMetricCounter(MetricCounter*& outMetric, MetricMetadata metadata);
```

Comment:
Why not return the pointer directly? This combined with a templated `AddMetric` would yield the following code:

```cpp
const auto *metric = AddMetric<MetricGauge>(metadata);
PPX_ASSERT_MSG(metric != nullptr, "Failed to create metric blabla");
```

From the tests I see errors are sometimes used (like ERROR_DUPLICATE_ELEMENT), but also assert are used, like `PPX_ASSERT_MSG(!metadata.name.empty(), "The metric name must not be empty");`.
Is there a real value to add error messages vs asserts?

What I mean is if we provide an API, the user interacts with at runtime, sure, we don't want to assert and crash because they gave the wrong name. But as metrics seem to be hardcoded, wouldn't a simple clean assert be simpler to debug when developing the benchmark?

Reply:
Agreed, I changed it for both metrics and runs creation. I added unit tests to check that the asserts trigger.